apps/backend/app/services/model_training/feature_engineer.py

```
@@ -664,4 +664,81 @@ def _apply_generic(
+            if pattern.lower() in code_lower:
+                raise ValueError(
+                    f"Feature code contains forbidden operation: {pattern}. "
+                    "Only pandas/numpy operations on dataframe columns are allowed."
+                )
+
+        logger.info("Feature code validation passed (basic checks only)")
+
+    async def apply_stored_feature(
+        self,
+        df: pd.DataFrame,
+        feature: 'StoredFeature'
```

Ruff (0.14.10)

Undefined name `StoredFeature`

(F821)